bytecomp/matching.ml

```diff
@@ -613,6 +591,8 @@ end = struct
 
   let to_pattern p = p
 
+  let head p = fst (Pattern_head.deconstruct p)
```

Comment:
Can you justify a bit the choice of keeping `pattern` equal to `Typedtree.pattern` and calling `Pattern_head.deconstruct` each time one asks for the head?

I had made the opposite choice in my branch (which wasn't necessarily better), so I'm curious about your rational.

Reply:
My reasoning was: `head * pattern list` and `pattern` are isomorphic, and the one on the right is simpler, so choosing it would give a simpler implementation.

I'm also a bit torn on these questions. If there was a strong performance benefit in going either way, that would force a choice. But I don't want to think of "getting the head of a pattern" as a costly operation, it should be a view that is essentially free. So I would rather be in favor of letting ourselves call `deconstruct` generously; if no performance issues can be measured, all is good. If we see a noticeable slowdown, we can try to make it faster, or we can revert on that choice -- the `pattern` type being abstract it's easy to change in retrospect.